training/deepspeech_training/train.py

```diff
@@ -75,145 +64,101 @@ def create_overlapping_windows(batch_x):
+        num_channels = Config.n_input
+        identity = np.eye(window_width * num_channels).reshape(
+            window_width, num_channels, window_width * num_channels
+        )
+        self.identity_filter = tf.constant(identity, tf.float32)
+        self.reshape_layer = tf.keras.layers.Reshape((-1, window_width * num_channels))
 
-        output, output_state = fw_cell(inputs=x,
-                                       dtype=tf.float32,
-                                       sequence_length=seq_length,
-                                       initial_state=previous_state)
+    def call(self, x):
+        x = tf.nn.conv1d(
+            input=x, filters=self.identity_filter, stride=1, padding="SAME"
+        )
+        return self.reshape_layer(x)
 
-    return output, output_state
+    def reshape(self, x):
+        return self.reshape_layer(x)
 
 
-def rnn_impl_cudnn_rnn(x, seq_length, previous_state, _):
-    assert previous_state is None # 'Passing previous state not supported with CuDNN backend'
+class Dense(tf.keras.layers.Dense):
+    def __init__(self, units, dropout_rate, layer_norm, **kwargs):
+        super().__init__(units, **kwargs)
 
-    # Hack: CudnnLSTM works similarly to Keras layers in that when you instantiate
-    # the object it creates the variables, and then you just call it several times
-    # to enable variable re-use. Because all of our code is structure in an old
-    # school TensorFlow structure where you can just call tf.get_variable again with
-    # reuse=True to reuse variables, we can't easily make use of the object oriented
-    # way CudnnLSTM is implemented, so we save a singleton instance in the function,
-    # emulating a static function variable.
-    if not rnn_impl_cudnn_rnn.cell:
-        # Forward direction cell:
-        fw_cell = tf.contrib.cudnn_rnn.CudnnLSTM(num_layers=1,
-                                                 num_units=Config.n_cell_dim,
-                                                 input_mode='linear_input',
-                                                 direction='unidirectional',
-                                                 dtype=tf.float32)
-        rnn_impl_cudnn_rnn.cell = fw_cell
+        self.dropout = tf.keras.layers.Dropout(dropout_rate) if dropout_rate != 0. else tf.keras.layers.Lambda(lambda x: x)
+        self.layer_norm = tf.keras.layers.LayerNormalization() if layer_norm else tf.keras.layers.Lambda(lambda x: x)
 
-    output, output_state = rnn_impl_cudnn_rnn.cell(inputs=x,
-                                                   sequence_lengths=seq_length)
+    def call(self, x):
+        return self.dropout(self.layer_norm(super().call(x)))
 
-    return output, output_state
 
-rnn_impl_cudnn_rnn.cell = None
+class Model(tf.keras.Model):
+    def __init__(self, dropout_rates=None, overlap=True, unroll=False, masking=True):
+        super().__init__(name='DeepSpeech_v{}'.format(int(file_relative_read('GRAPH_VERSION').strip())))
```

Comment:
is this some metadata we can poke? add more infos into?

Reply:
No, just a name scope that gets added to all the variables, sadly. There's some theoretical metadata entry points in the SavedModel format but given our export and native client setup it's a can of worms I didn't get into for now.